### PR TITLE
DNM crush: misc device class fixes

### PR DESCRIPTION
**Pull request by xiexingguo**
This patch set shall fix http://tracker.ceph.com/issues/20845
but there are other problems...

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas There are more problems that I am still working on...  do not try to test this yet...

And this one depends on https://github.com/ceph/ceph/pull/16680

**Comment by liewegas** [comment]
I like this https://github.com/ceph/ceph/pull/16657 better for the initial fix. And yeah, more needs fixing here.

I haven't looked past the remove_root problem, but I'm wondering if we can simplify life by disallowing 'link' from creating multiple parents for a single bucket?  If we only allow multiple parents for devices that may make life easier.  (And I think the multiple-parent thing is pretty uncharted territory otherwise!)

**Comment by xiexingguo** [comment]
> but I'm wondering if we can simplify life by disallowing 'link' from creating multiple parents for a single bucket?

We'd better not (for now). It can get people annoyed if we silently change the behaviour of existing CLIs.

**Comment by xiexingguo** [comment]
I am including kefu's fix(https://github.com/ceph/ceph/pull/16680) into this patch set.

@tchaikov Can you help to rerun the test scripts that are cause noises and see if this works?

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please...

**Comment by tchaikov** [comment]
@xiexingguo what do you mean by "rerun the test scripts"?

**Comment by xiexingguo** [comment]
Sorry, kefu. I should make myself more clear.
Generally, any qa test cases including `qa/workunits/mon/crush_ops.sh` will do.
For example:
http://pulpito.ceph.com/sage-2017-08-01_15:32:10-rados-wip-sage-testing-distro-basic-smithi/1469118/
http://pulpito.ceph.com/sage-2017-08-01_15:32:10-rados-wip-sage-testing-distro-basic-smithi/1469289/

Run this through a complete rados QA test suit will be better(there are possible other bugs `qa/workunits/mon/crush_ops.sh` might not cover).

**Comment by xiexingguo** [comment]
@tchaikov Another example: http://qa-proxy.ceph.com/teuthology/kchai-2017-07-31_14:22:05-rados-wip-kefu-testing-distro-basic-mira/1465207/teuthology.log

**Comment by tchaikov** [comment]
@xiexingguo see http://pulpito.ceph.com/kchai-2017-08-02_05:17:58-rados-wip-16698-kefu-distro-basic-smithi/

but it would be ideal if you could [apply](http://wiki.sepia.ceph.com/doku.php?id=gettingstarted) for the access to the sepia lab, so you can run the qa test by yourself without being blocked by me or other developers.

**Comment by xiexingguo** [comment]
> @xiexingguo see http://pulpito.ceph.com/kchai-2017-08-02_05:17:58-rados-wip-16698-kefu-distro-basic-smithi/

Thanks a lot, @tchaikov. 

> but it would be ideal if you could apply for the access to the sepia lab, so you can run the qa test by yourself without being blocked by me or other developers.

yeah, will do.

**Comment by xiexingguo** [comment]
The test passed and I believe this is ready for QA.

File: src/crush/CrushWrapper.cc [hunk 1]
```diff
@@ -491,6 +491,69 @@ bool CrushWrapper::_bucket_is_in_use(int item)
   return false;
 }
 
+int CrushWrapper::_get_rules_with_class(list<pair<int, list<string>>> *rules)
```
Comment (liewegas) [inline src/crush/CrushWrapper.cc]: Uh oh... we have bigger problems.  The bucket id of the shadow trees cannot change or else data gets reshuffled.  Looking at rebuild_roots_with_classes(), that's exactly what it does (potentially renumber all shadow buckets). So... that's what needs to be fixed.  (And assuming we fix that, this problem of fixing the rules should go away.)

Let me take a closer look at this this afternoon.  :(

**Comment by xiexingguo** [comment]
> The bucket id of the shadow trees cannot change

sadly I forget that...

**Comment by liewegas** [comment]
will push a branch in a few minutes with my changes

**Comment by xiexingguo** [comment]
In favor of https://github.com/ceph/ceph/pull/16805, which is the correct fix

**Pull request closed without merging**